// react-bootstrap/lib/Button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _elementType = _interopRequireDefault(require("prop-types-extra/lib/elementType"));

var _ThemeProvider = require("./ThemeProvider");

var _SafeAnchor = _interopRequireDefault(require("./SafeAnchor"));

var Button =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(Button, _React$Component);

  function Button() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = Button.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        bsPrefix = _this$props.bsPrefix,
        variant = _this$props.variant,
        size = _this$props.size,
        active = _this$props.active,
        className = _this$props.className,
        block = _this$props.block,
        type = _this$props.type,
        as = _this$props.as,
        innerRef = _this$props.innerRef,
        props = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["bsPrefix", "variant", "size", "active", "className", "block", "type", "as", "innerRef"]);
    var classes = (0, _classnames.default)(className, bsPrefix, active && 'active', bsPrefix + "-" + variant, block && bsPrefix + "-block", size && bsPrefix + "-" + size);

    if (props.href) {
      return _react.default.createElement(_SafeAnchor.default, (0, _extends2.default)({}, props, {
        as: as,
        innerRef: innerRef,
        className: (0, _classnames.default)(classes, props.disabled && 'disabled')
      }));
    }

    var Component = as || 'button';
    if (innerRef) props.ref = innerRef;
    return _react.default.createElement(Component, (0, _extends2.default)({}, props, {
      type: type,
      className: classes
    }));
  };

  return Button;
}(_react.default.Component);

Button.propTypes = {
  /**
   * @default 'btn'
   */
  bsPrefix: _propTypes.default.string,

  /**
   * One or more button variant combinations
   *
   * buttons may be one of a variety of visual variants such as:
   *
   * `'primary', 'secondary', 'success', 'danger', 'warning', 'info', 'dark', 'light', 'link'`
   *
   * as well as "outline" versions (prefixed by 'outline-*')
   *
   * `'outline-primary', 'outline-secondary', 'outline-success', 'outline-danger', 'outline-warning', 'outline-info', 'outline-dark', 'outline-light'`
   */
  variant: _propTypes.default.string,

  /**
   * Specifies a large or small button.
   *
   * @type ('sm'|'lg')
   */
  size: _propTypes.default.string,

  /** Spans the full width of the Button parent */
  block: _propTypes.default.bool,

  /** Manually set the visual state of the button to `:active` */
  active: _propTypes.default.bool,

  /**
   * Disables the Button, preventing mouse events,
   * even if the underlying component is an `<a>` element
   */
  disabled: _propTypes.default.bool,

  /** Providing a `href` will render an `<a>` element, _styled_ as a button. */
  href: _propTypes.default.string,

  /**
   * Defines HTML button type attribute.
   *
   * @default 'button'
   */
  type: _propTypes.default.oneOf(['button', 'reset', 'submit', null]),
  as: _elementType.default
};
Button.defaultProps = {
  variant: 'primary',
  active: false,
  disabled: false,
  type: 'button'
};

var _default = (0, _ThemeProvider.createBootstrapComponent)(Button, {
  prefix: 'btn',
  forwardRefAs: 'innerRef'
});

exports.default = _default;
module.exports = exports["default"];
