// react-bootstrap/lib/FormCheck.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ThemeProvider = require("./ThemeProvider");

var _FormContext = _interopRequireDefault(require("./FormContext"));

var _Feedback = _interopRequireDefault(require("./Feedback"));

var _FormCheckInput = _interopRequireDefault(require("./FormCheckInput"));

var _FormCheckLabel = _interopRequireDefault(require("./FormCheckLabel"));

var FormCheck =
/*#__PURE__*/
function (_React$Component) {
  (0, _inheritsLoose2.default)(FormCheck, _React$Component);

  function FormCheck() {
    return _React$Component.apply(this, arguments) || this;
  }

  var _proto = FormCheck.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        id = _this$props.id,
        bsPrefix = _this$props.bsPrefix,
        inline = _this$props.inline,
        disabled = _this$props.disabled,
        isValid = _this$props.isValid,
        isInvalid = _this$props.isInvalid,
        feedback = _this$props.feedback,
        inputRef = _this$props.inputRef,
        className = _this$props.className,
        style = _this$props.style,
        title = _this$props.title,
        type = _this$props.type,
        label = _this$props.label,
        children = _this$props.children,
        props = (0, _objectWithoutPropertiesLoose2.default)(_this$props, ["id", "bsPrefix", "inline", "disabled", "isValid", "isInvalid", "feedback", "inputRef", "className", "style", "title", "type", "label", "children"]);
    var hasLabel = label != null && label !== false && !children;

    var input = _react.default.createElement(_FormCheckInput.default, (0, _extends2.default)({}, props, {
      type: type,
      ref: inputRef,
      isValid: isValid,
      isInvalid: isInvalid,
      isStatic: !hasLabel,
      disabled: disabled
    }));

    return _react.default.createElement(_FormContext.default.Transform, {
      mapToValue: function mapToValue(_ref) {
        var controlId = _ref.controlId;
        return {
          controlId: id || controlId
        };
      }
    }, _react.default.createElement("div", {
      style: style,
      className: (0, _classnames.default)(className, bsPrefix, inline && bsPrefix + "-inline")
    }, children || _react.default.createElement(_react.default.Fragment, null, input, hasLabel && _react.default.createElement(_FormCheckLabel.default, {
      title: title
    }, label), (isValid || isInvalid) && _react.default.createElement(_Feedback.default, {
      type: isValid ? 'valid' : 'invalid'
    }, feedback))));
  };

  return FormCheck;
}(_react.default.Component);

FormCheck.propTypes = {
  /**
   * @default 'form-check'
   */
  bsPrefix: _propTypes.default.string,

  /**
   * The FormCheck `ref` will be forwarded to the underlying input element,
   * which means it will be a DOM node, when resolved.
   *
   * @type {ReactRef}
   * @alias {ref}
   */
  inputRef: _propTypes.default.any,

  /** A HTML id attribute, necessary for proper form accessibility. */
  id: _propTypes.default.string,

  /**
   * Provide a function child to manually handle the layout of the FormCheck's inner components.
   *
   * ````
   * <FormCheck>
   *   <FormCheck.Input isInvalid type={radio} />
   *   <FormCheck.Label>Allow us to contact you?</FormCheck.Label>
   *   <Feedback type="invalid">Yo this is required</Feedback>
   * </FormCheck>
   * ```
   */
  children: _propTypes.default.node,
  inline: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  title: _propTypes.default.string,
  label: _propTypes.default.node,

  /** The type of checkable. */
  type: _propTypes.default.oneOf(['radio', 'checkbox']).isRequired,

  /** Manually style the input as valid */
  isValid: _propTypes.default.bool.isRequired,

  /** Manually style the input as invalid */
  isInvalid: _propTypes.default.bool.isRequired,

  /** A message to display when the input is in a validation state */
  feedback: _propTypes.default.node
};
FormCheck.defaultProps = {
  type: 'checkbox',
  inline: false,
  disabled: false,
  isValid: false,
  isInvalid: false,
  title: ''
};
var DecoratedFormCheck = (0, _ThemeProvider.createBootstrapComponent)(FormCheck, {
  forwardRefAs: 'inputRef',
  prefix: 'form-check'
});
DecoratedFormCheck.Input = _FormCheckInput.default;
DecoratedFormCheck.Label = _FormCheckLabel.default;
var _default = DecoratedFormCheck;
exports.default = _default;
module.exports = exports["default"];
